### PR TITLE
feat(rust): paginate help texts

### PR DESCRIPTION
**Pull request by i-b-o-t**
using $PAGER if set, defaulting to `less` and falling back to `more` or no pagination if no pager is available.

Addresses #3434

## Checks

- [X] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [X] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [X] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [X] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

### CODE REVIEW

**Comment by mrinalwadhwa** [comment]
@i-b-o-t thank you for spending time on this.
Use of less worked like a charm, nice!

A few quick things:
1. Could you run `cargo fmt` and `cargo clippy` and fix any problems they report
2. Our commit messages are all lower case so please amend the commit message to `feat(rust): paginate help texts`

After that, I think we can merge this.

**Comment by i-b-o-t** [comment]
Sure, I can do that.

Also, I noticed that the exit status of the process should be either 0 or 2 depending on whether the help text is requested by the user explicitly or shown as a consequence of a usage error. By the same logic, in case no pagination is available, the output should go to `stdout` or `stderr` streams respectively.

Also, checking whether the pager is `less` should better be done via `path::file_name` in case the environment variable contains a full path.

So give me a minute to take care of that.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: Could have gone `unwrap_or_default().to_str().unwrap_or_default()` but wasn't sure what the preference was. Though technically the slightly more complicated syntax, I found it read more obvious at first glance this way.

**Comment by i-b-o-t** [comment]
In individual commits for easy reviewing. The first one is still the same, except for rewording its log entry to lowercase. Rebasing with `--autosquash` will turn them all into a single pretty one automagically, as you most probably know.

**Comment by i-b-o-t** [comment]
```shell
$ PAGER=false ./target/debug/ockam -h 1>/dev/null && echo OK
OK
$ PAGER=false ./target/debug/ockam 2>/dev/null || echo OK
OK
$ cat mock/swallow 
#!/bin/sh

$ PAGER=./mock/swallow ./target/debug/ockam -h && echo OK
OK
$ PAGER=./mock/swallow ./target/debug/ockam || echo OK
OK
$ cat mock/less 
#!/bin/sh

if test "${LESS}" = "-F"; then
	echo OK
fi
$ PAGER=./mock/less ./target/debug/ockam 
OK
```

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: You should use the error codes provided at [util/exitcode.rs](https://github.com/build-trust/ockam/blob/d59ccf0c41bc49d733fcc3733daadc3f9af34c8d/implementations/rust/ockam/ockam_command/src/util/exitcode.rs#L3)

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs:477]: I would rename the `help` variable to `error`. It was a little confusing reading this variable until I got here.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: Is this really needed? I thought both streams were flushed automatically when exiting the program

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: It would be very helpful to have the `if` block as a variable with a little explanation of what it's happening there (i.e. why an USAGE error if the error is written to stderr?)

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs:325]: Given that the `show_help` function is pretty short and it's only called here, I think it would be better to remove it and paste the code here.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: The following two functions could be simplified. There is a library called [`which`](https://crates.io/crates/which) you could use to get the binaries paths.

```rust
fn paginate(text: &StyledStr) -> Result<()> {
    let default_pager = get_env_with_default("PAGER", "less".to_string())?;
    let mut pagers = vec![&default_pager, "more", "bat"];
    for pager in pagers.drain(..) {
        if let Ok(path) = which::which(pager) {
            let path = path
                .to_str()
                .expect(&format!("{pager}'s path is not valid UTF-8"));
            if let Ok(()) = paginate_with(path, text) {
                break;
            }
        }
    }
    Ok(())
}

fn paginate_with(pager_path: &str, text: &StyledStr) -> Result<()> {
    let mut cmd = process::Command::new(pager_path);
    // Add needed arguments to the pager command.
    match pager_path {
        "less" => {
            // No pagination if the text fits entirely into the window.
            // Using env var in case a lesser `less` poses as `less`.
            cmd.env("LESS", "-F");
        }
        _ => {}
    }
    let mut child = cmd.stdin(process::Stdio::piped()).spawn()?;
    // Write text to pager's stdin.
    {
        let mut stdin = child.stdin.take().unwrap();
        stdin.write_all(format!("{}", text).as_bytes())?;
    }
    child.wait()?;
    Ok(())
}
```

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: Although this works as expected, the default behavior now strips the colors from the output. We still need that. I've tried passing the `-R` argument but it didn't work, so perhaps it's related to how the stdout/stderr streams are set up?

**Review by adrianbenavides — COMMENTED** [review]
Nice job @i-b-o-t! It's pretty much there but I've left some comments with potential improvements. Thanks for taking the time with this 🙏

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: It's taken from the `clap_builder::util::safe_exit` function and I just copied it over. I think you're right in case the compiler is instructed to perform a proper unwind (is that always the case for Ockam?). If not, I suspect it may depend on whether there's additional (non-OS) buffering in place.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: I basically like the suggestion. However, those are copied from `clap::Error::exit` which uses the same exit code for `USAGE` as prominent UNIX commands such as `ls` and the like. Using those constants would be a change in behavior, which I feel is outside the scope of this submission.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs:477]: I chose the name intentionally to indicate that, although the parameter's type is `Error`, it represents any help texts, including those that are issued upon the user's request - without an actual error!

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: I'm not a big fan of comments TBH and I thought I wrote too many of them already :sweat_smile: . Again, the code is copied from `clap::Error::exit`. But I see what you mean, arguably `clap::Error::use_stderr()` would better be called something like `represents_usage_error()`. I will introduce a boolean variable to "say it with code".

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs:325]: Humm... Sorry, no. I disagree again: `show_help` contains implementation details of an abnormal path of execution and the `run` function is the entry point. I'd prefer to give whoever is tracing into it a clean first impression by keeping it concise and on a single level of abstraction.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: How would that be a simplification? Also adding another dependency.. But more importantly, I think you are misreading my code, for this breaks when `PAGER=/opt/not-on-PATH/pager`. I added a fallback to `more` with devices in mind and don't think the overkill of probing any pager on the planet is necessary.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: I just tried `pgrep less` when running `git log` and it appears `less` is not running. Then again, pressing `h` shows a screen titled "SUMMARY OF LESS COMMANDS"...

Mysteriously
```shell
$ strings $( which git ) |grep LESS
LESS=FRX LV=-c
```

UPDATE: Ah, OK! Tthe `LV=-c` part is unrelated, a dash doesn't seem to be required, and the `-X` option seems a good idea too since it just keeps everything as-is on exit, so you can scroll back up to the help text (like to look up things while editing the command line).

**Comment by i-b-o-t** [comment]
:heart_decoration: Thanks for your review! Please don't get too negative of an impression from me disagreeing in many places and let's not obscure the fact that it *did* raise valid points.

**Comment by i-b-o-t** [comment]
Behaves just like Git now, AFAICT rebased and ready.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: Just to give you some more context, the error codes we use are the ones defined at https://man.freebsd.org/cgi/man.cgi?query=sysexits&apropos=0&sektion=0&manpath=FreeBSD+4.3-RELEASE&format=html

We'll review it again in case we should change them.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: Yeah, I get your point. The reason we try to make things as explicit and documented as possible is because this code (`ockam_command`) is read by people that are contributing to an opensource project for their first time (students, or people playing around with rust, for example), so we try to keep the code as less cryptic as possible, especially around non-obvious logic/assumptions. Of course, we don't want to document code that is self explanatory (as you said, sometimes a good variable name is good enough to explain what's going on).

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs:495]: The whole function feels obscure and with lots of hidden subtleties around the pager parsing. Here are some questions that I had while reading this particular block:

- This feels hacky, why not use an Option to represent "false" instead?
- If a user sets the "PAGER" env var to "false", why would they want to set it in the first place? I'd rather provide another env var to disable this feature.
- The user could have set the `NO_COLOR` env var, which is not taken into consideration here
- What's happening in this block? I had to read the documentation of the `with_cmd` and `exit` commands to learn that it's an early exit of the process. I also had to get into the details of the `paginate_with` to understand that they depend on each other. I wonder if this block should live in the `paginate_with` function.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs:514]: I'm not sure how portable would this be across different OS's. I still vote for using `which`. It will also look cleaner and it lets you check if the binary actually exists in the system.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs:530]: Ok, so this env var is only for internal use right?

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/lib.rs:544]: I'm sure we can do this without having to rerun the same command. Did you need it to add the `ENV_FORCE_COLOR`?

**Comment by adrianbenavides** [comment]
> 💟 Thanks for your review! Please don't get too negative of an impression from me disagreeing in many places and let's not obscure the fact that it did raise valid points.

No problem! That's the whole point of doing code reviews right? We are here to learn from each other 👍

**Comment by adrianbenavides** [comment]
Hey @i-b-o-t , I was taking another look at the PR and I just wanted to clarify something. The PR looks great, it works as expected and we could merge it as is. All my last comments are just questions that would help me understand better your solution and small details that could be improved, but they won't block the PR from merging.

That being said, feel free to mark the comments as resolved. We can tackle whatever is left in a separate PR.

And thanks again for taking the time with this issue, it ended up being more tricky than I initially anticipated!

**Comment by i-b-o-t** [comment]
Hey @adrianbenavides , thanks for giving it another chance - being able to take a step back and reconsider one's own position shows character! Defending it comes with wholehearted commitment, I know, nothing wrong with that as long as we realize when to hold on and take a break, a deep breath, another look or perspective into consideration.

**The simplest solutions is not always immediately obvious.**

I do think that Occam's Razor is a fine analogy to problem solving in software development and *given the scope and constants that constrain us here and now*, the implementation in a nutshell is as simple as it gets.

Once we widen the scope (e.g. by introducing further command line options, changing the public interface, as we both have suggested above) or find a way to redefine the constants (e.g. by requesting additional Clap API), there is always room for improvement piecing the puzzle together in a coherent fashion. However, in order for these efforts to be both manageable and motivating, I think it is a good idea to chop things up into individual working steps and then working that plan without too much deviation. Of course there are exceptions, like something being completely infeasible, but we don't have that case here. Also, there is almost always some nits to pick on; doesn't mean we should just because we could (invert priorities).

> The PR looks great, it works as expected and we could merge it as is

Just pushed the squashed commit.

EDIT: To complete my philosophical sermon and add another quote

> That's the whole point of doing code reviews right? We are here to learn from each other

I agree, plus hopefully merging better code than we would without.

From that angle, nitpicking certainly has its place in reviews, as it can mean teaching each other (often ideally generalizable) details to be aware of, but it also bears the risk of wasting time and inverting priorities, so there's a pair of anti-forces that need to balanced.

**Comment by adrianbenavides** [comment]
Thanks for the message @i-b-o-t , I really appreciate it 🙏 . One last thing: we require the [commits to be signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits). Let me know if you want me to sign it or you want to do it yourself. It's a one time operation and it's pretty quick to set up.

**Comment by i-b-o-t** [comment]
> One last thing: we require the [commits to be signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).

Are they not? I set up PGP the other day. I need to check on it...

```
$ git config commit.gpgsign
true
```

Shouldn't I see an error if they were not. Oh, or does it also need that "Signed off by" line in the commit message?

**Comment by adrianbenavides** [comment]
> Shouldn't I see an error if they were not. Oh, or does it also need that "Signed off by" line in the commit message?

That's not needed. Looks like your commit is signed but can't be verified. I think you have to upload your public key to your githu account for the verification to work.

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: I already had a previous state return those, but then just ended up using `clap::Error::exit` again for simplicity.

Since you mentioned elsewhere that you see a lot of novice contributors, probably also people who are new to Rust (I include myself in that group), and there's quite some setup to be done in total to get on board, the remapping the of these exit codes could be a really nice and easy follow-up task to file. Basically:

```
clap_error.print();

// from `clap_builder::util::safe_exit` (not needed if we always
// unwind the stack on exit):
let _ = std::io::stdout().lock().flush();
let _ = std::io::stderr().lock().flush();

let is_usage_error = clap_error.use_stderr();
// might as well replicate(private) `clap::Error::stream()`
// making it more obvious why that is so

process::exit(if is_usage_error { USAGE } else { OK });
```

File: implementations/rust/ockam/ockam_command/src/lib.rs
Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs:544]: Where's my reply? Possibly hit the wrong button - gotta get used to GitHub again. Why the re-run is the way of the least resistance was posted elsewhere.

Introducing a command line flag like GCC's `-fdiagnostics-color=[auto|never|always]` setting `clap::ColorChoice` (which happens to be an `enum` over the same options) would be a possible way to get rid of that internal environment variable in a follow-up task.

Comment (i-b-o-t) [reply implementations/rust/ockam/ockam_command/src/lib.rs:544]: That said, I personally don't really really like the above suggestion, since some settings already look crappy using a pager by default. Git doesn't need it. And there is means to control the automatic by connecting the output channels to (possibly pseudo-) terminals or not.

You mentioned a conventional `NO_COLOR` env var. Supporting that one (and keeping the internal one or piggy-backing on `NO_COLOR` with an internal, "impossible"-to-collide value that instead forces colors to ON - just for communicating with the child - both cases being mutually exclusive) would be another low-hanging fruit and better aligned with the rest of the existing interface.

Comment (adrianbenavides) [reply implementations/rust/ockam/ockam_command/src/lib.rs:544]: We already support the `NO_COLOR` env var and we have a way to determine if the streams support color, although we should introduce the internal "force-color" setting as you mentioned. There's definitely room for improvement on this regard. We'll give it a thought after testing this for a while.

**Review by adrianbenavides — APPROVED** [review]
Let's get this merged! Thanks again @i-b-o-t, this was a great contribution! 🚀

**Pull request merged**
